### PR TITLE
Windows msvc

### PR DESCRIPTION
**Pull request by Rantanen**
- Instead of disabling Luv/Busted, we can fix the build by specifying LUAJIT in cmake files for MSVC
- The gperf issue can be fixed by demanding MSVC to use functions for ceil/floor instead of inline intrinsics
- Also tweaked the build_msvc.bat a bit:
  - Skipped gperf download if it already exists.
  - Enabled specifying cmake options (such as PYTHON_EXECUTABLE) from command line through the bat.
  - Forced GYP (libuv) to use VS2015.
  - ... and switched CMake to use VS2015 generator instead of NMake.

That last one was just personal preference. Felt more in control with solution files that could be opened in Visual Studio if needed.

Feel free to omit stuff/cherry pick changes you want.

### CODE REVIEW

File: .ci/build_msvc.bat [hunk 1]
```diff
@@ -1,19 +1,34 @@
 
 call "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat" amd64
-:: For gperf
-C:\msys64\usr\bin\bash -lc "pacman --verbose --noconfirm -Su" || goto :error
-C:\msys64\usr\bin\bash -lc "pacman --verbose --noconfirm --needed -S gperf" || goto :error
 
+set CMAKE_GENERATOR="Visual Studio 14 2015 Win64"
```
Comment (equalsraf) [inline .ci/build_msvc.bat]: My olny reason for using nmake is that it is usually more verbose on errors. With the VS generator the vcvarsall line can be removed, I think.

File: src/nvim/eval.c [hunk 2]
```diff
@@ -6706,6 +6706,11 @@ static int get_env_tv(char_u **arg, typval_T *rettv, int evaluate)
 }
 
 #ifdef INCLUDE_GENERATED_DECLARATIONS
+// MSVC uses intrinsic inline implementation for ceil and floor by default.
+// funcs.generated.h requires function forms for function pointers.
+# ifdef _MSC_VER
+#  pragma function(ceil, floor)
+# endif
```
Comment (equalsraf) [inline src/nvim/eval.c]: @Rantanen nice work cathing this. Can this be in win_defs.h instead?

**Comment by equalsraf** [comment]
Awesome, thanks.

**Comment by equalsraf** [comment]
@Rantanen I've added most of your changes, also enabled the functional tests. Some tests are failing, this seems to be caused by the missing libintl.

Log: https://ci.appveyor.com/project/equalsraf/neovim/build/1373/job/avdu5wkw4hvo2b1v
Source: https://github.com/neovim/neovim/blob/51937e1322de0e4348c217947c117f54f3c664c3/src/nvim/ex_docmd.c#L141

**Comment by Rantanen** [comment]
Oh, the tests are working? That's neat! Last I checked nvim.exe wouldn't even start so I guess this might mean it might.

I tested switching the `#pragma` to `win_defs.h` and that seemed to work fine.

Also rebased some of the remaining changes to the `build_msvc.bat`. These aren't needed for the CI builds, but would help if the batch file is being used by other developers when building on Windows.

Tell me if you want to keep the `build_msvc.bat` more deterministic and aimed purely for the CI-environment and I'll figure out some other way to allow easy dev-builds with the MSVC toolchain. Currently I've been using the `build_msvc.bat` for this. :)

**Comment by equalsraf** [comment]
> Oh, the tests are working? That's neat! Last I checked nvim.exe wouldn't even start so I guess this might mean it might.

A lot more tests are failing in debug builds
- https://ci.appveyor.com/project/equalsraf/neovim/build/1374/job/m36fsroanp6psa0t

most of them seem to be shada related.

> Tell me if you want to keep the build_msvc.bat more deterministic and aimed purely for the CI

Here its fine. For merging upstream its probably best to leave it out, since it is in a separate commit this is easy to do in a rebase.

I would leave out the gperf check, we need to do some CI fixes about dependenceies anyway, so this can be done later on.

**Comment by Rantanen** [comment]
> I would leave out the gperf check, we need to do some CI fixes about dependenceies anyway, so this can be done later on.

I'll just keep it in my branch then as I don't have msys installed locally. So don't be surprised if it pops up somewhere else as well - just ignore it when picking commits. :)

**Comment by equalsraf** [comment]
> I'll just keep it in my branch then as I don't have msys installed locally.

In Unix third-party builds gperf. In Windows with MSVC we could download a binay build instead.

**Comment by Rantanen** [comment]
> we could download a binary build instead.

That makes sense.

**Comment by equalsraf** [comment]
Fixed gperf recipe to grab binary builds from gnuwin32

**Comment by Rantanen** [comment]
Tried looking into the libintl/iconv problem today. Given those are C libraries, we should be able to crosslink MinGW built libraries into MSVC build. Unfortunately I couldn't find 64-bit prebuilt binaries. The Gnuwin32 binaries are built for x86.

**Comment by Rantanen** [comment]
Oh, also closing this pull request.

All the changes that were included here are now either taken into the windows-msvc branch or implemented some other way.

**Pull request closed without merging**